app/src/main/java/com/github/swent/echo/data/repository/RepositoryImpl.kt

```diff
@@ -1,65 +1,209 @@
-        return remoteDataSource.joinEvent(userId, event.eventId)
+        if (isOffline()) {
+            throw RepositoryStoreWhileNoInternetException("Event Join")
+        }
+        val remoteResult = remoteDataSource.joinEvent(userId, event.eventId)
+        localDataSource.joinEvent(userId, event.eventId)
+        return remoteResult
     }
 
     override suspend fun leaveEvent(userId: String, event: Event): Boolean {
-        return remoteDataSource.leaveEvent(userId, event.eventId)
+        if (isOffline()) {
+            throw RepositoryStoreWhileNoInternetException("Event Join")
+        }
+        val remoteResult = remoteDataSource.leaveEvent(userId, event.eventId)
+        localDataSource.leaveEvent(userId, event.eventId)
+        return remoteResult
     }
 
     override suspend fun getJoinedEvents(userId: String): List<Event> {
-        return remoteDataSource.getJoinedEvents(userId)
+        if (
+            isOffline() ||
+                notExpired(events_last_cached_joined.getOrDefault(userId, 0), EVENT_CACHE_TTL)
+        ) {
+            return localDataSource.getJoinedEvents(userId, FETCH_ALL)
+        }
+        val nonExpired = localDataSource.getJoinedEvents(userId, EVENT_CACHE_TTL)
+        val nonExpiredIds = nonExpired.map { it.eventId }
+        val remoteUpdate = remoteDataSource.getJoinedEventsNotIn(userId, nonExpiredIds)
+        localDataSource.leaveEventsNotIn(userId, nonExpiredIds)
+        localDataSource.joinEvents(userId, remoteUpdate.map { it.eventId })
+        events_last_cached_joined.put(userId, currentTimeStamp())
+        return nonExpired + remoteUpdate
     }
 
     override suspend fun getTag(tagId: String): Tag? {
-        return remoteDataSource.getTag(tagId)
+        if (isOffline()) {
+            return localDataSource.getTag(tagId, FETCH_ALL)
+        }
+        val localResult = localDataSource.getTag(tagId, TAG_CACHE_TTL)
+        if (localResult == null) {
+            val remoteResult = remoteDataSource.getTag(tagId)
+            if (remoteResult != null) localDataSource.setTag(remoteResult)
+            return remoteResult
+        }
+        return localResult
     }
 
     override suspend fun getSubTags(tagId: String): List<Tag> {
-        return remoteDataSource.getSubTags(tagId)
+        if (
+            isOffline() ||
+                notExpired(tags_last_cached_all, TAG_CACHE_TTL) ||
+                notExpired(tags_last_cached_subs.getOrDefault(tagId, 0), TAG_CACHE_TTL)
+        ) {
+            return localDataSource.getSubTags(tagId, FETCH_ALL)
+        }
+        val nonExpired = localDataSource.getSubTags(tagId, TAG_CACHE_TTL)
+        val nonExpiredIds = nonExpired.map { it.tagId }
+        val remoteUpdate = remoteDataSource.getSubTagsNotIn(tagId, nonExpiredIds)
+        localDataSource.deleteSubTagsNotIn(tagId, nonExpiredIds)
+        localDataSource.setTags(remoteUpdate)
+        tags_last_cached_subs.put(tagId, currentTimeStamp())
```

Comment:
Nitpicking: could also be converted to an assignment.

Reply:
same as in the above comment^